File: Day5/day5.js
const inputLines = readLineByLine('input.txt');


class Procedure {
    constructor(txt) {
        const splitted = txt.split(" ");
        this.move = parseInt(splitted[1]);
        this.from = parseInt(splitted[3]) - 1;
        this.to = parseInt(splitted[5]) - 1;
    }
}

class Crates {
    constructor(size) {
        this.crates = [];
        for (let i = 0; i < size; i++) {
            this.crates[i] = [];
        }
    }

    copy() {
        const crates = new Crates(this.crates.length);
        for (let i = 0; i < this.crates.length; i++) {
            crates.crates[i] = [];
            for (let j = 0; j < this.crates[i].length; j++) {
                crates.crates[i].push(this.crates[i][j]);
            }
        }
        return crates;
    }

    fillCrates(line) {
        let j = 0;
        for (let i = 1; i < line.length; i += 4) {
            if (line[i] != " ") this.crates[j].unshift(line[i]);
            j++;
        }
    }

    applyProcedure(proc, is9001 = false) {
        const start = this.crates[proc.from].length - proc.move;
        const end = this.crates[proc.from].length;
        const sublist = this.crates[proc.from].slice(start, end);
        Array.prototype.push.apply(this.crates[proc.to], is9001 ? sublist : sublist.reverse());
        this.crates[proc.from].splice(start, end);
    }

    get topOfStacks() { return this.crates.map((e) => e[e.length - 1]).join(""); }
}


const procedures = [];
const crates = new Crates(9);

for (let i = 0; i < inputLines.length; i++) {
    const line = inputLines[i];

    if (i < 8) {
        crates.fillCrates(line);
        continue;
    }

    if (i > 9) procedures.push(new Procedure(line));
}

const crates9001 = crates.copy();

for (const proc of procedures) {
    crates.applyProcedure(proc);
    crates9001.applyProcedure(proc, true);
}

console.log(`[Part1] Top of each stack CrateMover9000: ${crates.topOfStacks}`);
console.log(`[Part2] Top of each stack CrateMover9001: ${crates9001.topOfStacks}`);

function readLineByLine(filePath) {
    const fs = require('fs');
    const allFileContents = fs.readFileSync(filePath, 'utf-8');
    const inputLines = allFileContents.split(/\r?\n/);

    return inputLines;
}

File: Day5/day5.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func main() {
	inputLines := readLineByLine("input.txt")

	crates := newCrates(9)
	procs := []*Procedure{}

	for i, line := range inputLines {
		if i < 8 {
			crates.fillCrates(line)
			continue
		}

		if i > 9 {
			procs = append(procs, procFromStr(line))
		}
	}

	crates9001 := crates.copy()

	for _, proc := range procs {
		crates.applyProcedure(proc, false)
		crates9001.applyProcedure(proc, true)
	}

	fmt.Println("[Part1] Top of each stack CrateMover9000:", crates.topOfStacks())
	fmt.Println("[Part2] Top of each stack CrateMover9001:", crates9001.topOfStacks())
}

type Crates struct {
	crates [][]string
}

func newCrates(size int) *Crates {
	return &Crates{
		make([][]string, size),
	}
}

func (c *Crates) copy() *Crates {
	crates := newCrates(len(c.crates))
	copy(crates.crates, c.crates)
	return crates
}

func (c *Crates) fillCrates(line string) {
	j := 0
	for i := 1; i < len(line); i += 4 {
		char := string(line[i])
		if char != " " {
			c.crates[j] = append([]string{char}, c.crates[j]...)
		}
		j++
	}
}

func (c *Crates) applyProcedure(proc *Procedure, is9001 bool) {
	end := len(c.crates[proc.from])
	start := end - proc.move

	sublist := c.crates[proc.from][start:end]
	if !is9001 {
		sort.SliceStable(sublist, func(i, j int) bool {
			return i > j
		})
	}
	c.crates[proc.to] = append(c.crates[proc.to], sublist...)
	c.crates[proc.from] = c.crates[proc.from][:start]
}

func (c *Crates) topOfStacks() string {
	txt := ""
	for _, crate := range c.crates {
		txt += crate[len(crate)-1]
	}
	return txt
}

type Procedure struct {
	move int
	from int
	to   int
}

func procFromStr(txt string) *Procedure {
	splitted := strings.Split(txt, " ")

	move, _ := strconv.Atoi(splitted[1])
	from, _ := strconv.Atoi(splitted[3])
	to, _ := strconv.Atoi(splitted[5])

	return &Procedure{move, from - 1, to - 1}
}

func readLineByLine(filePath string) []string {
	readFile, err := os.Open(filePath)

	if err != nil {
		fmt.Println(err)
	}
	fileScanner := bufio.NewScanner(readFile)
	fileScanner.Split(bufio.ScanLines)

	fileLines := []string{}
	for fileScanner.Scan() {
		fileLines = append(fileLines, fileScanner.Text())
	}

	readFile.Close()

	return fileLines
}
